### ISSUE TITLE
Generic way to handle per-project workspace configuration

### ISSUE THREAD

**Issue by merrickluo** [emacs-lsp/lsp-mode#3049]
Is your feature related or already mentioned on [the wishlist](https://github.com/emacs-lsp/lsp-mode/discussions/2892)?
Seems not.

`lsp-register-custom-settings` is great for setting global configurations for language servers. But if users want to customize some settings base on project, it requires the specific client implementation like making the variable buffer-local or completely overrides the `workspace/configuration` response like in [lsp-eslint.el](https://github.com/emacs-lsp/lsp-mode/blob/d904e8a4472b537f0bff1315a61fe3cf3b5ebcf5/clients/lsp-eslint.el#L250).

eglot provides a generic buffer-local variable to [set workspace configuration](https://github.com/joaotavora/eglot#per-project-server-configuration) but lacks global settings. 

So how about adding a similar buffer-local variable, and merge it to the global settings set by `lsp-register-custom-settings` before respond to `workspace/configuration`, maybe also add a new hook for `make-lsp-client` so clients can handle this more easily.

**Comment by yyoncho** [comment]
You  can use dir locals for per project configuration.

**Comment by ericdallo** [comment]
we can probably add a example on the docs of using the dir locals, PR welcome

**Comment by merrickluo** [comment]
I do know about dir locals and this can be done right now. say that I want this [`completionBudget`](https://github.com/golang/tools/blob/master/gopls/doc/settings.md#completionbudget-timeduration) setting for gopls which does not preset in `lsp-go.el`. This is what I think I have to add to my config.
```emacs-lisp
(defvar-local lsp-go-completion-budget "100ms")
(put 'lsp-go-completion-budget 'safe-local-variable 'stringp)
(lsp-register-custom-settings '(("gopls.completionBudget" lsp-go-completion-budget)))
```

then in dir-locals.el
```emacs-lisp
((go-mode . ((lsp-go-completion-budget . "200ms"))))
```

Also, most of the settings are defined with `defcustom`, in order to use them in dir locals, I have to use `(make-variable-buffer-local 'variable)` and `(put 'variable 'safe-local-variable)`.

In this example, I think it should be easier for users to just add
```emacs-lisp
 (go-mode
  . ((lsp-workspace-configuration
      . ((:gopls . (:completionBudget "200ms")))))))
```

I might be wrong, if there is already an easier way to do this, please let me know.

**Comment by yyoncho** [comment]
@merrickluo typically, the defcustom is part of lsp-mode gopls package so you will only need the part in the .dir-local file. The eglot solution is problematic because if you want to change 10 settings for all projects and one property for each project you will have to duplicate the 10 settings for each project.

**Comment by merrickluo** [comment]
Yeah, maybe I didn't make it clear in the op, I did say that the eglot lacks a way of setting global configuration, I'm not suggesting completely switch to that, but combine them. It will be much easier for users to overriding part of the configuration.
